lib/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp

```diff
@@ -1773,56 +1805,51 @@ void ExternalGenericMetadataBuilderContext<Runtime>::populateMachOSymbols(
     }
   }
 
-  // Find all the symbols.
-  for (auto &sym : file->objectFile->symbols()) {
-    auto type = sym.getType();
-    if (!type)
-      continue;
-    if (type.get() != llvm::object::SymbolRef::ST_Data &&
-        type.get() != llvm::object::SymbolRef::ST_Function)
-      continue;
-
-    auto name = sym.getName();
-    if (!name) {
-      consumeError(name.takeError());
-      continue;
+  // Find all the segments.
+  for (auto &command : file->objectFile->load_commands()) {
+    auto addSegment = [&](const auto &segmentCommand) {
+      Segment segment{segmentCommand.segname, segmentCommand.vmaddr,
+                      segmentCommand.vmsize};
+      if (segment.name == "__TEXT")
+        file->baseAddress = segment.address;
+      file->segments.push_back(segment);
+    };
+    if (command.C.cmd == llvm::MachO::LC_SEGMENT_64) {
+      auto segmentCommand = file->objectFile->getSegment64LoadCommand(command);
+      addSegment(segmentCommand);
+    } else if (command.C.cmd == llvm::MachO::LC_SEGMENT) {
+      auto segmentCommand = file->objectFile->getSegmentLoadCommand(command);
+      addSegment(segmentCommand);
     }
+  }
 
-    auto flags = sym.getFlags();
-    if (!flags) {
-      consumeError(flags.takeError());
+  // Find all the exported symbols.
+  llvm::Error err = llvm::Error::success();
+  for (auto &ex : file->objectFile->exports(err)) {
+    // Just in case there's somehow a symbol with a name that's not valid UTF-8,
+    // ignore it, since we can't represent it in our JSON output.
+    if (!llvm::json::isUTF8(ex.name()))
```

Comment:
We should be able to represent non-UTF-8 things in the JSON output; they just need to be escaped. JSON doesn't require the strings it holds to be UTF-8… it just requires that the JSON _itself_ be UTF-8.

Whether LLVM's JSON code deals with that is, of course, another question entirely.

Reply:
I added this check after some trash got into my names (it's a fun game figuring out the lifetime of the storage referred to by any given `llvm::StringRef`) and caused the JSON serializer to assert, so, yeah.

Reply:
Figures.